### ISSUE TITLE
reproduce `git clone -b` with go-git

### ISSUE THREAD

**Issue by caarlos0** [go-git/go-git#97]
with `git`, you can `git clone -b` and pass either a branch or a tag, and it will work.

passing `refs/heads/$BRANCH` or `refs/tags/$TAG`, but I wanted to pass a single thing regardless it being a branch or tag (e.g.: I don't know if it is a branch or a tag at clone time for some reason).

is it possible to do the same with go-git?

**Comment by caarlos0** [comment]
OK, digging into git source code, found this:

https://github.com/git/git/blob/master/builtin/clone.c#L550-L568

then I looked into `find_ref_by_name` implementation:

https://github.com/git/git/blob/master/remote.c#L1312-L1318

after digging a bit more, found a way to implement it, which I'm not sure is correct (and probably isn't the best way either):

```go
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"

	"github.com/go-git/go-billy/v5/osfs"
	"github.com/go-git/go-git/v5"
	"github.com/go-git/go-git/v5/config"
	"github.com/go-git/go-git/v5/plumbing"
	"github.com/go-git/go-git/v5/plumbing/cache"
	"github.com/go-git/go-git/v5/plumbing/transport/http"
	"github.com/go-git/go-git/v5/storage/filesystem"
)

func main() {
	var repo = flag.String("repo", "https://github.com/caarlos0-graveyard/test", "repo")
	var branch = flag.String("branch", "master", "branch")
	flag.Parse()

	directory, err := ioutil.TempDir("", "")
	handlerr(err)
	fmt.Println(directory)
	defer os.RemoveAll(directory)

	var auth = &http.BasicAuth{
		Username: os.Getenv("GITHUB_TOKEN"),
	}

	var remote = git.NewRemote(
		filesystem.NewStorage(
			osfs.New(directory),
			cache.NewObjectLRUDefault(),
		),
		&config.RemoteConfig{
			URLs: []string{*repo},
		},
	)

	refs, err := remote.List(&git.ListOptions{
		Auth: auth,
	})
	handlerr(err)

	var ref = plumbing.NewBranchReferenceName(*branch)
	if !exists(ref, refs) {
		ref = plumbing.NewTagReferenceName(*branch)
		if !exists(ref, refs) {
			log.Fatalln(*branch + " not found")
		}
	}

	fmt.Println("Using", ref)

	r, err := git.PlainClone(directory, false, &git.CloneOptions{
		URL:               *repo,
		Auth:              auth,
		SingleBranch:      true,
		Depth:             1,
		RecurseSubmodules: git.DefaultSubmoduleRecursionDepth,
		Progress:          os.Stdout,
		ReferenceName:     ref,
	})
	handlerr(err)

	head, err := r.Head()
	handlerr(err)

	fmt.Println(head.Strings())
}

func exists(ref plumbing.ReferenceName, refs []*plumbing.Reference) bool {
	for _, ref2 := range refs {
		if ref.String() == ref2.Name().String() {
			return true
		}
	}
	return false
}

func handlerr(err error) {
	if err != nil {
		log.Fatalln(err)
	}
}

```


can someone help me validate this?

**Comment by AriehSchneier** [comment]
Fixed by #755

**Comment by caarlos0** [comment]
Thanks
